sdk/internal/perf/implementation.go

```diff
@@ -4,295 +4,36 @@
-				}
-
-				warmUpPerSecondCount = append(warmUpPerSecondCount, thisSecondCount)
-				warmUpLastPrint += 1.0
-				if int(warmUpLastPrint) == warmUpDuration {
-					// We can have odd scenarios where we send this
-					// message, and the final message below right after
-					warmUpLastPrint += 1.0
-				}
-			}
-		}
-
-		thisSecondCount := warmupCount - sumInts(warmUpPerSecondCount)
-		c <- runResult{
-			warmup:        true,
-			completed:     true,
-			count:         thisSecondCount,
-			parallelIndex: index,
-			timeInSeconds: time.Since(warmUpStart).Seconds(),
-		}
-	}
-
-	timeStart := time.Now()
-	totalCount := 0
-	lastPrint := 1.0
-	perSecondCount := make([]int, 0)
-	for time.Since(timeStart).Seconds() < float64(duration) {
-		err := p.Run(context.Background())
-		if err != nil {
-			c <- runResult{err: err}
-		}
-		totalCount += 1
-
-		// Every second (roughly) we send an update through the channel
-		if time.Since(timeStart).Seconds() > lastPrint {
-			thisCount := totalCount - sumInts(perSecondCount)
-			c <- runResult{
-				count:         thisCount,
-				parallelIndex: index,
-				completed:     false,
-				timeInSeconds: time.Since(timeStart).Seconds(),
-			}
-			lastPrint += 1.0
-			perSecondCount = append(perSecondCount, thisCount)
-
-			if int(lastPrint) == duration {
-				// if we are w/in one second of the end time, we do not
-				// want to send any more results, we'll just send a final result
-				lastPrint += 10.0
-			}
-		}
-	}
-
-	elapsed := time.Since(timeStart).Seconds()
-	lastSecondCount := totalCount - sumInts(perSecondCount)
-	c <- runResult{
-		completed:     true,
-		count:         lastSecondCount,
-		parallelIndex: index,
-		timeInSeconds: elapsed,
-	}
-
-	if testProxyURLs != "" {
-		// Stop the proxy now
-		err := proxyTransportsSuite[ID].stop()
-		if err != nil {
-			c <- runResult{err: err}
-		}
-		proxyTransportsSuite[ID].SetMode("live")
-	}
-}
-
-// runCleanup takes care of the semantics for tearing down a single iteration of a performance test.
-func runCleanup(p PerfTest) error {
-	return p.Cleanup(context.Background())
-}
-
-// runResult is the result sent back through the channel for updates and final results
-type runResult struct {
-	// number of iterations completed since the previous message
-	count int
-
-	// The time the update comes from
-	timeInSeconds float64
-
-	// if there is an error it will be here
-	err error
-
-	// true when this is the last result from a go routine
-	completed bool
 
-	// Index of the goroutine
-	parallelIndex int
+	// wg is used to keep track of the number of goroutines created
+	wg sync.WaitGroup
 
-	// indicates if result is from a warmup start
-	warmup bool
-}
+	// number of processes to use, the --maxprocs flag
+	numProcesses int
+)
 
-// parse the TestProxy input into a slice of strings
+// parseProxyURLs splits the --test-proxy input with the delimiter ';'
 func parseProxyURLS() []string {
-	var ret []string
 	if testProxyURLs == "" {
-		return ret
+		return []string{}
```

Comment:
Generally, `nil` should be the same as empty slice. Would that work here? I think it's more idiomatic.